<a href="https://colab.research.google.com/github/ncsu-mskohli/dsa405-project/blob/main/notebooks/DSA405_002_FA26_P2_mskohli.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. The Audit

In [ ]:
# ---------------------------------------------------------------------------
# P2 setup
# ---------------------------------------------------------------------------

import pandas as pd, numpy as np

DATA = "https://raw.githubusercontent.com/ncsu-mskohli/dsa405-project/main/data/raw/"
df = pd.read_csv(DATA + "leith_used_cars.csv")

In [ ]:
# CLEANING CODEBLOCK 1

# The original dataset, referenced as `df` in this file...
# ...has the information of multiple variables...
# ...contained in just one: "Make & Model"...
# ...and so we must first separate them.

# note that pandas automatically removed the first row with the headers, the csv had 1992 rows

pattern_condition = r'^(Used|Certified Pre-Owned)'
df['Condition'] = df['Make & Model'].str.extract(pattern_condition, expand=False)

temp_series = df['Make & Model'].str.replace(pattern_condition, '', regex=True).str.strip()

df['Year'] = temp_series.str.extract(r'^(\d{4})', expand=False)

df['Make'] = temp_series.str.extract(r'^\d{4}\s+(\w+)', expand=False)

df['Model'] = temp_series.str.replace(r'^\d{4}\s+\w+\s+', '', regex=True).str.strip()

df_c = df[['Condition', 'Year', 'Make', 'Model', 'Price']].copy()

# Printing a sample of our reorganized dataframe, `df_c`
firstotherCondition = (df_c[df_c['Condition'] == 'Certified Pre-Owned'].index[0])
print(df_c.head(1 + firstotherCondition + 1))


              Condition  Year        Make                  Model   Price
0                  Used  2025  Volkswagen            Taos 1.5T S  $24998
1                  Used  2025  Volkswagen          Taos 1.5T SEL  $29582
2                  Used  2026        Jeep        Wrangler Sahara  $44594
3                  Used  2021         BMW           X3 xDrive30i  $31899
4                  Used  2023       LEXUS                 ES 250  $36693
5                  Used  2023        Jeep  Grand Cherokee Summit  $39233
6                  Used  2019         BMW     M850i M850i xDrive  $44197
7                  Used  2025         BMW                   430i  $47335
8                  Used  2024       LEXUS    RX 350 Premium Plus  $56257
9                  Used  2025  Volkswagen            Taos 1.5T S  $25772
10                 Used  2025  Volkswagen     Taos 1.5T SE Black  $26960
11  Certified Pre-Owned  2025       Honda                CR-V LX  $28948
12                 Used  2026      Nissan       Rog

In [ ]:
# Dataframe info
df_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1991 entries, 0 to 1990
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Condition  1991 non-null   object
 1   Year       1991 non-null   object
 2   Make       1991 non-null   object
 3   Model      1991 non-null   object
 4   Price      1991 non-null   object
dtypes: object(5)
memory usage: 77.9+ KB


In [ ]:
# Auditing the `Condition` column

# Reading it in as the dtype it should be
Condition_audit = (df_c.Condition).astype('string')

# Checking for null values
print(f"Null values in the original: {df_c.Condition.isna().sum()}")
print(f"Null values after reading it as `string`: {Condition_audit.isna().sum()}")
print()

# The number of distinct values and...
#...outputting their full set if it is relevant and reasonable.
print(f"Distinct values in the original: {df_c.Condition.nunique()}")
print(f"Distinct values after reading it in as strings: {Condition_audit.nunique()}")
print()
print(Condition_audit.unique())
print()

# Frequency of values
print(Condition_audit.value_counts())

Null values in the original: 0
Null values after reading it as `string`: 0

Distinct values in the original: 2
Distinct values after reading it in as strings: 2

<StringArray>
['Used', 'Certified Pre-Owned']
Length: 2, dtype: string

Condition
Used                   1625
Certified Pre-Owned     366
Name: count, dtype: Int64


In [ ]:
# Auditing the `Year` column

# Reading it in as the dtype it should be
Year_audit = pd.to_numeric(df_c['Year'], errors='coerce')

# Checking for null values
print(f"Null values in the original: {df_c.Year.isna().sum()}")
print(f"Null values after reading it as numeric: {Year_audit.isna().sum()}")
print()

# The number of distinct values and...
#...outputting their full set if it is relevant and reasonable.
print(f"Distinct values in the original: {df_c.Year.nunique()}")
print(f"Distinct values after reading it in as numeric: {Year_audit.nunique()}")
print()
print("Sorted full set of values, ascending:")
print(np.sort(Year_audit.unique()))
print()

# Max, min, range
Year_max = Year_audit.max()
Year_min = Year_audit.min()
Year_range = Year_max - Year_min
print(f"Max: {Year_max} Min: {Year_min} Range: {Year_range}")
print()

# The frequency of each value
print(Year_audit.value_counts(dropna=False))

Null values in the original: 0
Null values after reading it as numeric: 0

Distinct values in the original: 24
Distinct values after reading it in as numeric: 24

Sorted full set of values, ascending:
[2002 2003 2004 2005 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017
 2018 2019 2020 2021 2022 2023 2024 2025 2026 2027]

Max: 2027 Min: 2002 Range: 25

Year
2024    433
2025    380
2023    284
2026    192
2022    158
2021    126
2020     98
2019     86
2018     54
2017     42
2016     36
2015     30
2014     19
2013     18
2012     10
2010      6
2009      5
2011      4
2003      4
2008      2
2004      1
2002      1
2005      1
2027      1
Name: count, dtype: int64


In [ ]:
# Auditing the `Make` column

# Reading it in as the dtype it should be
Make_audit = (df_c.Make).astype('string')

# Checking for null values
print(f"Null values in the original: {df_c.Make.isna().sum()}")
print(f"Null values after reading it as `string`: {Make_audit.isna().sum()}")
print()

# The number of distinct values and...
#...outputting their full set if it is relevant and reasonable.
print(f"Distinct values in the original: {df_c.Make.nunique()}")
print(f"Distinct values after reading it in as strings: {Make_audit.nunique()}")
print()
print(Make_audit.unique())
print()

# Frequency of top values
print(Make_audit.value_counts().head(10))

Null values in the original: 0
Null values after reading it as `string`: 0

Distinct values in the original: 39
Distinct values after reading it in as strings: 39

<StringArray>
['Volkswagen',       'Jeep',        'BMW',      'LEXUS',      'Honda',
     'Nissan',      'Acura',   'Mercedes',  'Chevrolet',       'Ford',
     'Subaru',   'Chrysler',     'Toyota',       'Audi',        'Kia',
   'Maserati',      'Volvo',    'Porsche',    'Hyundai',        'GMC',
      'Dodge',       'MINI',    'Lincoln',    'Vinfast',        'Ram',
      'Buick',      'Mazda',   'INFINITI',    'Genesis',      'Tesla',
   'CADILLAC',       'Alfa',     'Rivian',    'McLaren',       'FIAT',
      'Scion', 'Mitsubishi',   'Cadillac',      'Aston']
Length: 39, dtype: string

Make
Honda         304
Toyota        203
Jeep          148
Volkswagen    142
Ford          117
Audi          109
Mercedes      108
Nissan        102
Chevrolet      95
BMW            85
Name: count, dtype: Int64


In [ ]:
# Auditing the `Model` column

# Reading it in as the dtype it should be
Model_audit = (df_c.Model).astype('string')

# Checking for null values
print(f"Null values in the original: {df_c.Model.isna().sum()}")
print(f"Null values after reading it as `string`: {Model_audit.isna().sum()}")
print()

# The number of distinct values and...
#...outputting their full set if it is relevant and reasonable.
print(f"Distinct values in the original: {df_c.Model.nunique()}")
print(f"Distinct values after reading it in as strings: {Model_audit.nunique()}")
print()

# Frequency of top values
print(Model_audit.value_counts().head(20))

# Checking for whitespace causing duplicates
print()
Model_audit_nunique_beforewhitespacestrip = Model_audit.nunique()
Model_audit_nunique_afterwhitespacestrip = Model_audit.str.strip().nunique()
assert Model_audit_nunique_beforewhitespacestrip == Model_audit_nunique_afterwhitespacestrip

Null values in the original: 0
Null values after reading it as `string`: 0

Distinct values in the original: 993
Distinct values after reading it in as strings: 993

Model
Civic Sport            27
RAV4 XLE               24
CR-V EX-L              21
CR-V Hybrid Sport-L    20
Rogue SV               18
Tiguan 2.0T SE         17
Camry SE               17
CR-V EX                15
Corolla LE             15
HR-V Sport             13
Odyssey EX-L           12
Sentra SV              12
Pilot Touring          11
Nautilus Reserve       11
HR-V LX                11
Sportage EX            10
Explorer XLT           10
Taos 1.5T SE           10
Accord EX              10
X5 xDrive40i            9
Name: count, dtype: Int64



In [ ]:
# Auditing the `Price` column

# Reading it in as the dtype it should be
Price_audit = pd.to_numeric(df_c['Price'], errors='coerce')

# Checking for null values
print(f"Null values in the original: {df_c.Price.isna().sum()}")
print(f"Null values after reading it as numeric: {Price_audit.isna().sum()}")
print()

# Checking for null values after stripping the `$`
Price_audit = df_c['Price'].str.replace(r'[\$,]', '', regex=True)
Price_audit = pd.to_numeric(Price_audit, errors='coerce')
print(f"Null values after stripping the `$`: {Price_audit.isna().sum()}")
print(f"Rate (rounded to 1 decimal place): {Price_audit.isna().sum() / len(Price_audit) * 100:.1f}%")
print()

# The number of distinct values and...
#...outputting their full set if it is relevant and reasonable.
print(f"Distinct values in the original: {df_c.Price.nunique(dropna=False)}")
print(f"Distinct values after reading it in as numeric: {Price_audit.nunique(dropna=False)}")
print(f"Distinct values after reading it in as numeric, non-NaN: {Price_audit.nunique()}")
print()

# Max, min, and range
Price_max = Price_audit.max()
Price_min = Price_audit.min()
Price_range = Price_max - Price_min
print(f"Max: {Price_max} Min: {Price_min} Range: {Price_range}")
print()

print(df_c.iloc[Price_audit[Price_audit == Price_max].index[0]])
print()
print(df_c.iloc[Price_audit[Price_audit == Price_min].index[0]])
print()

# Frequency, checking for duplicates due to distinct values not being as high as expected
print(Price_audit.value_counts().head(20))
print()
print(df_c[Price_audit == 19998.0])

Null values in the original: 0
Null values after reading it as numeric: 1991

Null values after stripping the `$`: 38
Rate (rounded to 1 decimal place): 1.9%

Distinct values in the original: 1514
Distinct values after reading it in as numeric: 1514
Distinct values after reading it in as numeric, non-NaN: 1513

Max: 300683.0 Min: 4988.0 Range: 295695.0

Condition    Certified Pre-Owned
Year                        2024
Make                     Porsche
Model                  911 Turbo
Price                    $300683
Name: 82, dtype: object

Condition        Used
Year             2010
Make            Honda
Model        Civic LX
Price           $4988
Name: 1776, dtype: object

Price
19998.0    10
21678.0     9
24998.0     8
30678.0     8
34997.0     7
24997.0     7
28197.0     7
29997.0     7
29998.0     7
31197.0     7
10678.0     6
40678.0     6
27678.0     6
29678.0     6
39997.0     6
16678.0     6
24597.0     5
26688.0     5
25548.0     5
24678.0     5
Name: count, dtype: int64

    

Most classes of defect were found to not be present by the basic functions of the audit.

For columns `Condition` and `Make`, there were no null values and the number of values so few that there not found to be any defects on manual inspection. This also verified for example that there were no total or header rows present.

For column `Year`, the minimum, maximum, and range found that all values were sensible, and there were no null values. For column `Price`, it was similar. To be extra sure, the rows for the maximum and minimum were checked and found to be appropriate. The column had to be stripped of its currency sign to check for null values properly after converting to numeric and some were found. Duplicates for `Price` were suspected due to `Price` usually being extremely unique and there being a surprising number of repeated values. However, upon inspection, there were numerous valid observations with the same values for `Price`.

No null values were found for `Model`. Manipulating and working with the data in the project in the future will reveal any errors or issues with this column, and is to some extent, part of the next stage of the project as model names being matched with the other data source and issues with that will be revealed then or remove themselves when no match is found. We tested removing whitespace and it did not affect the number of values or consistency of the data, so it was a class of defect not found.

# 2. The data dictionary

Note:
* Each observation is a used car for sale.
* In the audit, some variables were treated as strings but will be made categories during the clean.

Variables:

| Field | What goes in it |
|---|---|
| **Condition** | `Used` or `Certified Pre-Owned` |
| Type | category |
| Units | N.A. |
| Factor levels | `Used` or `Certified Pre-Owned` |
| Valid range | N.A. |
| Missingness | None |
| Notes | |

| | |
|---|---|
| **Year** | Year of manufacture of the car at sale |
| Type | int |
| Units | year, XXXX |
| Factor levels | N.A. |
| Valid range | positive, any reasonable year for cars being manufactured |
| Missingness | None |
| Notes | |

| | |
|---|---|
| **Make** | Make of the car at sale |
| Type | category |
| Units | car brands, "Make" |
| Factor levels | Too many to list here, but they were outputted in the audit. |
| Valid range | N.A. |
| Missingness | None |
| Notes | |

| | |
|---|---|
| **Model** | Model of the car at sale. |
| Type | string |
| Units | names of car models |
| Factor levels | N.A. |
| Valid range | N.A. |
| Missingness | None |
| Notes | It includes trims as well. This may not violate rules for tidy data as the intention is for the dataset it merges with to do the same and consider it as one specific variation of a model.|

| | |
|---|---|
| **Price** | Price of the car at sale. |
| Type | float |
| Units | dollars |
| Factor levels | N.A. |
| Valid range | Any positive number, reasonable to the human eye. |
| Missingness | 38 values, 1.9% - no relevant meaning; just no stated price |
| Notes | Includes a standard $698 "Doc Fee". |

# 3. The cleaning log
| # | Column | Change made | Rows/cells affected | Why | What is lost |
|---|---|---|---|---|---|
| 1 | `Make & Model` | Four new columns created from it in dataframe `df`, them and `Price` counted into the new dataframe `df_c` | 1991 rows | To separate variables stuck together in one column. | Nothing |
| 2 | `Condition` | Converting from `object` to `category` | 1991 cells | More appropriate data type, efficiency | Nothing |
| 3 | `Make` | Converting from `object` to `category` | 1991 cells | More appropriate data type, efficiency | Nothing |
| 4 | `Model` | Converting from `object` to `string` | 1991 cells | More appropriate data type, efficiency | Nothing |
| 5 | `Price` | Stripping `$` from values and then converting via `pd.to_numeric` | 1991 cells | Identify truly null values; More appropriate data type, efficiency | Nothing |
| 6 | `Price` | Removing rows which have a null value for this column | 38 rows | No longer useful for the dataset, data could not be repaired, it was simply missing | 38 rows |
| 7| `Year` | Converting via `pd.to_numeric` | 1953 cells | More appropriate data type, efficiency | Nothing |

In [ ]:
# CLEANING CODEBLOCK 2
# Reseting everything
%reset -f
# Making changes based on the audit
# All decisions recorded in the cleaning log

# Setup again
import pandas as pd, numpy as np

DATA = "https://raw.githubusercontent.com/ncsu-mskohli/dsa405-project/main/data/raw/"
df = pd.read_csv(DATA + "leith_used_cars.csv")

print("Info on the original dataframe")
print("-")
df.info(memory_usage='deep')
print()

# 1
pattern_condition = r'^(Used|Certified Pre-Owned)'
df['Condition'] = df['Make & Model'].str.extract(pattern_condition, expand=False)

temp_series = df['Make & Model'].str.replace(pattern_condition, '', regex=True).str.strip()

df['Year'] = temp_series.str.extract(r'^(\d{4})', expand=False)

df['Make'] = temp_series.str.extract(r'^\d{4}\s+(\w+)', expand=False)

df['Model'] = temp_series.str.replace(r'^\d{4}\s+\w+\s+', '', regex=True).str.strip()

df_c = df[['Condition', 'Year', 'Make', 'Model', 'Price']].copy()

print("Information after column reorganization prior to cleaning")
print("-")
df_c.info(memory_usage='deep')
print()

# 2
prior = df_c.Condition.nunique(dropna=False)
df_c['Condition'] = df_c['Condition'].astype('category')
after = df_c.Condition.nunique(dropna=False)
assert prior == after

# 3
prior = df_c.Make.nunique(dropna=False)
df_c['Make'] = df_c['Make'].astype('category')
after = df_c.Make.nunique(dropna=False)
assert prior == after

# 4
prior = df_c.Model.nunique(dropna=False)
df_c['Model'] = df_c['Model'].astype('string')
after = df_c.Model.nunique(dropna=False)
assert prior == after

# 5
Price_nulls = df_c.Price.isna().sum()
df_c['Price'] = df_c['Price'].str.replace(r'[\$,]', '', regex=True)
df_c['Price'] = pd.to_numeric(df_c['Price'], errors='coerce')
Price_nulls = df_c.Price.isna().sum()

# Added during step 6 as a pre-check, consider this "5.5"
# Was originally `assert` and not `assert not`
assert not len(df_c) == len(df_c.dropna())

# 6
priorRows = len(df_c)
df_c = df_c.dropna(subset=['Price'])
postRows = len(df_c)
assert df_c.Price.isna().sum() == 0
assert postRows == priorRows - Price_nulls

# 7
prior = df_c.Year.nunique(dropna=False)
df_c['Year'] = pd.to_numeric(df_c['Year'], errors='coerce')
after = df_c.Year.nunique(dropna=False)
assert prior == after

# This confirms no more nulls exist as len counts NaN values
assert len(df_c) == len(df_c.dropna())

# Info after cleaning
print("Information after cleaning")
print("-")
df_c.info(memory_usage='deep')
print()
print(df_c.nunique())

# Saving the cleaned dataset into the temporary local directory, "files"
df_c.to_csv('leith_used_cars_cleanedX.csv', index=False)


Info on the original dataframe
-
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1991 entries, 0 to 1990
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Make & Model  1991 non-null   object
 1   Price         1991 non-null   object
dtypes: object(2)
memory usage: 270.4 KB

Information after column reorganization prior to cleaning
-
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1991 entries, 0 to 1990
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Condition  1991 non-null   object
 1   Year       1991 non-null   object
 2   Make       1991 non-null   object
 3   Model      1991 non-null   object
 4   Price      1991 non-null   object
dtypes: object(5)
memory usage: 551.9 KB

Information after cleaning
-
<class 'pandas.core.frame.DataFrame'>
Index: 1953 entries, 0 to 1953
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype   

# 4. Row and column accounting


```
Rows in raw file                      1991
Rows removed as exact duplicates      0
Rows removed as near-duplicates       0
Rows removed for other reasons        38   (each one logged above)
Rows in cleaned file                  1953

Columns in raw file                   2
Columns dropped                       1   (redundant, we extracted new columns from it and copied what was left into a new column. it was not directly dropped, but a new dataframe replaced an old one and left it out)
Columns added                         4
Columns in cleaned file               5
```

# 5. The Provenance Brief
Leith Cars is a car dealership in Raleigh, NC. They maintain an inventory of used cars they have for sale on their website, `leithcars.com`.

Through a scrape of the website on 10/03/2026, we gathered the following information on each car: its `Condition`, whether it is simply `Used` or also `Certified Pre-Owned`; its `Year` of manufacture; its `Make`; its `Model`; and finally, and most importantly, its `Price`.

Mileage was not scraped although it was information given on the website. We will rely on the year of manufacture to serve its purpose as an indicator.

We then organized and cleaned this data to end up with a dataset that lists 1953 used cars listed for sale on October 3rd, 2026 that were manufactured between 2002 and 2027, of 39 makes, and 975 models.

The vast majority of these cars, nearly 80% of them, were manufactured between `2019` and `2025`. The makes are more evenly distributed. The largest share is of `Honda`, which accounts for 15% of the cars being sold.

We intend to use this dataset to help make comparisons between Raleigh's used car market with national data.

# AI use note
* I used AI to help me understand how to build the scrape and to help write code to fix errors and speeden up the process.
* I used AI to entirely rewrite the scrape to be done differently.
* I used AI to entirely write the code for extracting the new columns I wanted.
* I used AI to help correct syntax, identify the right keywords and also replicate code at times throughout the file for certain processes.

